pkg/storage/fs/ocis/tree.go

```diff
@@ -0,0 +1,263 @@
+package ocis
+
+import (
+	"context"
+	"encoding/hex"
+	"math/rand"
+	"net/url"
+	"os"
+	"path/filepath"
+	"time"
+
+	provider "github.com/cs3org/go-cs3apis/cs3/storage/provider/v1beta1"
+	"github.com/cs3org/reva/pkg/errtypes"
+	"github.com/cs3org/reva/pkg/user"
+	"github.com/google/uuid"
+	"github.com/pkg/errors"
+	"github.com/pkg/xattr"
+	"github.com/rs/zerolog/log"
+)
+
+// Tree manages a hierarchical tree
+type Tree struct {
+	pw            PathWrapper
+	DataDirectory string
+}
+
+// NewTree creates a new Tree instance
+func NewTree(pw PathWrapper, dataDirectory string) (TreePersistence, error) {
+	return &Tree{
+		pw:            pw,
+		DataDirectory: dataDirectory,
+	}, nil
+}
+
+// GetMD returns the metadata of a node in the tree
+func (t *Tree) GetMD(ctx context.Context, node *NodeInfo) (os.FileInfo, error) {
+	md, err := os.Stat(filepath.Join(t.DataDirectory, "nodes", node.ID))
+	if err != nil {
+		if os.IsNotExist(err) {
+			return nil, errtypes.NotFound(node.ID)
+		}
+		return nil, errors.Wrap(err, "tree: error stating "+node.ID)
+	}
+
+	return md, nil
+}
+
+// GetPathByID returns the fn pointed by the file id, without the internal namespace
+func (t *Tree) GetPathByID(ctx context.Context, id *provider.ResourceId) (relativeExternalPath string, err error) {
+	var node *NodeInfo
+	node, err = t.pw.NodeFromID(ctx, id)
+	if err != nil {
+		return
+	}
+
+	relativeExternalPath, err = t.pw.Path(ctx, node)
+	return
+}
+
+// CreateDir creates a new directory entry in the tree
+func (t *Tree) CreateDir(ctx context.Context, node *NodeInfo) (err error) {
+
+	// TODO always try to fill node?
+	if node.Exists || node.ID != "" { // child already exists
+		return
```

Comment:
return an error ? important for webdav's "405 method not allowed" on already existing collection

Reply:
ocs is making a stat call to return the proper codes. but we can return  `errtypes.AlreadyExists` here